### PR TITLE
Merge data_flow branch into master

### PR DESCRIPTION
**Pull request by sararunkel**
Separated out data_management scripts into new folder data_flow and refactored push_data.py and sync_field_data.py

### CODE REVIEW

**Comment by janineA** [comment]
@sararunkel I made a few minor changes to clarify comments and clean up formatting and committed them to the data_flow branch. There are a couple things I would like you to do:
- I tried to run the tests, cleaned up one error, but still see a couple errors when running from my MAC that seem to be related to hardcoded paths. For example, in test_distribute_data it expected /RAF_data/QAtools but got /Users/janine/dev/projects/CAESAR/QAtools. Could you look into this? If you need more context I can send you the actual errors in an email.
- Please add information on the purpose and usage of sync_field_data to the readme.md file.
Let me know when you are done and I will continue my code review.

In general everything looks great! This is a huge improvement to what we had and will make it so much easier to make updates when our needs change.

**Comment by sararunkel** [comment]
@janineA Sounds good! I will work on making sure the tests can be run from anywhere and adding documentation. I know that one failed because I just fixed the fact that the path was hardcoded in the code but forgot to fix it in the tests.

I'll add a section in the readme for sync_field_data as well and send you an email when it's ready for another look over.

File: scripts/data_flow/sync_field_data.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+import sys
+import smtplib
+from email.mime.text import MIMEText
+
+# set up variables
+# Have to get PROJ_DIR, project and aircraft from env here even though will
+# get them again when import fieldProc_setup, because need them to find
+# location of fieldPro_setup.
+
+full_proj_dir = os.environ['PROJ_DIR'] + '/' + os.environ['PROJECT'] + '/' + os.environ['AIRCRAFT']+'/'
```
Comment (janineA) [inline scripts/data_flow/sync_field_data.py:25]: @sararunkel Once you resolve how to gracefully catch if an env var isn't set in your test suite, please add that same error catching here. Right now, if AIRCRAFT isn't set, the code crashes:

>  $ ./sync_field_data.py 
> Traceback (most recent call last):
>   File "/Users/janine/dev/aircraft_projects/scripts/data_flow/./sync_field_data.py", line 25, in <module>
>     full_proj_dir = os.environ['PROJ_DIR'] + '/' + os.environ['PROJECT'] + '/' + os.environ['AIRCRAFT']+'/'
>                                                                                  ~~~~~~~~~~^^^^^^^^^^^^
>   File "<frozen os>", line 714, in __getitem__
> KeyError: 'AIRCRAFT'
>

File: scripts/data_flow/push_data.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
+from _setup import Setup #setup, myLogger
+import sys, os,glob
+import  _GDrive, _process,_NAS,_FTP,_zip
+sys.path.insert(0, os.environ['PROJ_DIR'] + '/' + os.environ['PROJECT'] + '/' + os.environ['AIRCRAFT'] + '/scripts')
```
Comment (janineA) [inline scripts/data_flow/push_data.py:4]: @sararunkel Once you resolve how to gracefully catch if an env var isn't set in your test suite, please add that same error catching here. Right now, if AIRCRAFT isn't set, the code crashes:

> Traceback (most recent call last):
>   File "/Users/janine/dev/aircraft_projects/scripts/data_flow/push_data.py", line 1, in <module>
>     from _setup import Setup #setup, myLogger
>     ^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/Users/janine/dev/aircraft_projects/scripts/data_flow/_setup.py", line 8, in <module>
>     sys.path.insert(0, os.environ['PROJ_DIR'] + '/' + os.environ['PROJECT'] + '/' + os.environ['AIRCRAFT'] + '/scripts')
>                                                                                     ~~~~~~~~~~^^^^^^^^^^^^
>   File "<frozen os>", line 714, in __getitem__
> KeyError: 'AIRCRAFT'

**Review by janineA — CHANGES_REQUESTED** [review]
I am not skilled with the GitHub pull request workflow. I can't seem to resolve any of the comments I have made that have already been addressed. Thinking that I might need to submit a review in order to be able to resolve things, I am going to try doing that now. Hopefully I don't mess everything up.

File: scripts/data_flow/readme.md [hunk 3]
```diff
@@ -0,0 +1,224 @@
+# PUSH_DATA Readme to understand data flow
+
```
[the remark this answers is not in the document]
Comment (sararunkel) [reply scripts/data_flow/readme.md]: @janineA Added table of contents!

File: scripts/data_flow/test/test__setup.py [hunk 4]
```diff
@@ -1,8 +1,19 @@
...
+except Exception as e:
+    # It's better to catch specific exceptions
+    print(e)
+    errorMsg = "Please set the environment variables PROJ_DIR, PROJECT, and AIRCRAFT to point to an existing project directory."
+    for var in ['PROJ_DIR', 'PROJECT', 'AIRCRAFT']:
+        if var in os.environ:
+            errorMsg += f"\n{var} is set to: {os.environ[var]}"
+        else:
+            errorMsg += f"\n{var} is not set."
+    sys.exit(errorMsg)
```
[the remark this answers is not in the document]
Comment (sararunkel) [reply scripts/data_flow/test/test__setup.py]: @janineA I couldn't replicate this error, but I added a new check, check_env.py, that checks if the environment variables are set, and then checks if the os.environ['PROJ_DIR'] + '/' + os.environ['PROJECT'] + '/' + os.environ['AIRCRAFT'] path exists and will give an error message if either of those fail. It worked for me, but let me know it is still not giving you a clean system exit.

File: scripts/data_flow/sync_field_data.py (hunk 1 above)
Comment (sararunkel) [reply scripts/data_flow/sync_field_data.py:25]: Added in the same check_env.py method to sync_field_data.py

File: scripts/data_flow/push_data.py (hunk 2 above)
Comment (sararunkel) [reply scripts/data_flow/push_data.py:7]: Added in the same check here

File: scripts/data_flow/test/test__setup.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (janineA) [reply scripts/data_flow/test/test__setup.py]: Looks like it's a Mac vs linux issue with sys.exit(). Your solution works on mercury but still doesn't work on my Mac. I did some googling and found that wrapping the sys.exit in a try-except seems to work:

        try:
            sys.exit(message)
        except SystemExit as e:
            print(e.code)
            os._exit(1) # Any number works here

From this page: https://stackoverflow.com/questions/173278/is-there-a-way-to-prevent-a-systemexit-exception-raised-from-sys-exit-from-bei
"If your problem is SystemExit being caught at outer levels (i.e., unittest), then be the outer level yourself! Wrap your main code in a try/except block, catch SystemExit, and call os._exit() there, and only there! "

**Review by janineA — APPROVED** [review]

**Comment by janineA** [comment]
@sararunkel Looks good. Go ahead and merge into master, pull master to eol-groundstation3, and  edit the crontab -e on eol-rosetta to the updated sync_field_data.py path.

Thanks!

**Pull request merged**